File: 2021/lib/day_01.rb
class Day01
  def part1(input)
    measurements = input.split.map(&:to_i)
    measurements[..-2].zip(measurements[1..]).count { |x, y| y > x }
  end

  def part2(input)
    measurements = input.split.map(&:to_i)
    summed = measurements[..-3].zip(measurements[1..-2], measurements[2..]).map(&:sum)
    summed[..-2].zip(summed[1..]).count { |x, y| y > x }
  end
end

File: 2025/lib/day_01.ex
defmodule Day01 do
  def part_1(input) do
    input
    |> String.split()
    |> Enum.map_reduce(50, fn rotation, pointing_at ->
      new_pointing_at = (pointing_at + rotation_amount(rotation)) |> Integer.mod(100)
      {new_pointing_at, new_pointing_at}
    end)
    |> elem(0)
    |> Enum.count(fn x -> x == 0 end)
  end

  def part_2(input) do
    input
    |> String.split()
    |> Enum.map_reduce(50, fn rotation, pointing_at ->
      rotation_amount = rotation_amount(rotation)
      complete_rotations = div(rotation_amount, 100) |> abs()
      new_pointing_at = pointing_at + rem(rotation_amount, 100)
      corrected_pointing_at = new_pointing_at |> Integer.mod(100)

      cond do
        corrected_pointing_at == 0 or
            (pointing_at != 0 and (new_pointing_at < 0 or new_pointing_at > 99)) ->
          {1 + complete_rotations, corrected_pointing_at}

        true ->
          {complete_rotations, corrected_pointing_at}
      end
    end)
    |> elem(0)
    |> Enum.sum()
  end

  defp rotation_amount("L" <> amount_str), do: -String.to_integer(amount_str)
  defp rotation_amount("R" <> amount_str), do: String.to_integer(amount_str)
end
